crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+
+        let Some(first) = definitions.next() else {
+            return DefinitionType::Type(Type::Unbound);
+        };
+
+        if let Some(second) = definitions.next() {
+            let context = self.typing_context();
+            let mut builder = UnionTypeBuilder::new(&context);
+            builder = builder.variant(first).variant(second);
+
+            for variant in definitions {
+                builder = builder.variant(variant);
+            }
+
+            // The fact that the interner is local to a body means that we can't reuse the same union type
+            // across different call sites. But that's something we aren't doing yet anyway. Our interner doesn't
+            // deduplicate union types that are identical.
+            DefinitionType::Union {
+                ty: builder.build(),
+            }
+        } else {
+            DefinitionType::Type(first)
+        }
+    }
+}
+
+enum DefinitionType {
+    Union { ty: UnionType },
+    Type(Type),
+}
+
+impl DefinitionType {
+    fn into_type(self, types: &mut TypeInferenceBuilder) -> Type {
+        match self {
+            DefinitionType::Union { ty } => types.union_ty(ty),
+            DefinitionType::Type(ty) => ty,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ruff_db::file_system::FileSystemPathBuf;
+    use ruff_db::vfs::system_path_to_file;
+
+    use crate::db::tests::TestDb;
+    use crate::module::resolver::{set_module_resolution_settings, ModuleResolutionSettings};
+    use crate::name::Name;
+    use crate::red_knot::types::{public_symbol_ty_by_name, Type, TypingContext};
+
+    fn setup_db() -> TestDb {
+        let mut db = TestDb::new();
+
+        set_module_resolution_settings(
+            &mut db,
+            ModuleResolutionSettings {
+                extra_paths: Vec::new(),
+                workspace_root: FileSystemPathBuf::from("/src"),
+                site_packages: None,
+                custom_typeshed: None,
+            },
+        );
+
+        db
+    }
+
+    fn assert_public_ty(db: &TestDb, file_name: &str, symbol_name: &str, expected: &str) {
+        let file = system_path_to_file(db, file_name).expect("Expected file to exist.");
+
+        let ty = public_symbol_ty_by_name(db, file, symbol_name).unwrap_or(Type::Unknown);
+        assert_eq!(ty.display(&TypingContext::global(db)).to_string(), expected);
+    }
+
+    #[test]
+    fn follow_import_to_class() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_files([
+            ("src/a.py", "from b import C as D; E = D"),
+            ("src/b.py", "class C: pass"),
+        ])?;
+
+        assert_public_ty(&db, "src/a.py", "E", "Literal[C]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn resolve_base_class_by_name() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_file(
+            "src/mod.py",
+            r#"
+class Base:
+    pass
+
+class Sub(Base):
+    pass"#,
```

Comment:
Per my follow-up comment on the symbol table PR, I'm not convinced by the text-offsets argument, and I'd still rather use `dedent` in tests (unless we don't need it? see below) instead of this style of indentation for multi-line code in tests.

Reply:
> I'm not convinced by the text-offsets argument,

Wait until you have to debug some more byte offset errors, but maybe it's different in the type checker context. Debugging byte offsets was something I did frequently when working on the formatter. 

I personally find both about equally ugly because rustfmt can't format them for you. But I'm not sure if I like having `dedent` calls in all tests.

Reply:
I wouldn't want dedent calls in every test. I'd want helpers for writing file contents in tests with less boilerplate, which would also call dedent on the given source.

And yeah, I think caring about byte offsets will be a lot less common in these tests than in parser or formatter tests.